src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2041,13 +2041,13 @@ protected override void AfterLeftChildHasBeenVisited(BoundBinaryOperator binary)
 
             BinaryOperatorKind op = binary.OperatorKind.Operator();
 
-            // learn from non-null constant
+            // learn from non-null expression
             BoundExpression operandComparedToNonNull = null;
-            if (isNonNullConstant(binary.Left))
+            if (leftType.IsNotNull && rightType.MayBeNull && binary.Right.ConstantValue?.IsNull != true)
             {
                 operandComparedToNonNull = binary.Right;
             }
-            else if (isNonNullConstant(binary.Right))
+            else if (leftType.MayBeNull && binary.Left.ConstantValue?.IsNull != true && rightType.IsNotNull)
```

Comment:
>rightType.IsNotNull [](start = 88, length = 19)

Please sort this condition so that it mirrors the one above.

Reply:
I simplified this code by checking for a null constant before looking for a NonNull to MaybeNull comparison.